pkg/reconciler/apibinding/apibinding_controller.go

```diff
@@ -0,0 +1,347 @@
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	apiExports, err := c.apiExportsIndexer.ByIndex(indexAPIExportsByAPIResourceSchema, key)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	for _, apiExport := range apiExports {
+		c.enqueueAPIExport(apiExport)
+	}
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s controller", controllerName)
+	defer klog.Infof("Shutting down %s controller", controllerName)
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	klog.Infof("processing key %q", key)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(ctx context.Context, key string) error {
+	namespace, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("invalid key: %q: %v", key, err)
+		return nil
+	}
+	clusterName, name := clusters.SplitClusterAwareKey(clusterAwareName)
+
+	obj, err := c.apiBindingsLister.Get(key) // TODO: clients need a way to scope down the lister per-cluster
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+	old := obj
+	obj = obj.DeepCopy()
+
+	if err := c.reconcile(ctx, obj); err != nil {
+		return err
+	}
+
+	// If the object being reconciled changed as a result, update it.
+	if !equality.Semantic.DeepEqual(old.Status, obj.Status) {
+		oldData, err := json.Marshal(apisv1alpha1.APIBinding{
+			Status: old.Status,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to Marshal old data for apibinding %s|%s/%s: %w", clusterName, namespace, name, err)
+		}
+
+		newData, err := json.Marshal(apisv1alpha1.APIBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				UID:             old.UID,
+				ResourceVersion: old.ResourceVersion,
+			}, // to ensure they appear in the patch as preconditions
+			Status: obj.Status,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to Marshal new data for apibinding %s|%s/%s: %w", clusterName, namespace, name, err)
+		}
+
+		patchBytes, err := jsonpatch.CreateMergePatch(oldData, newData)
```

Comment:
wondering, what does this do when removing an initializer? Will it merge empty list with non-empty list and hence, don't remove the initializer?

Reply:
```go
oldBinding := &apisv1alpha1.APIBinding{
		Status: apisv1alpha1.APIBindingStatus{
			Initializers: []string{"init"},
		},
	}

	newBinding := &apisv1alpha1.APIBinding{
		Status: apisv1alpha1.APIBindingStatus{
			Initializers: []string{},
		},
	}

	oldData, _ := json.Marshal(oldBinding)
	newData, _ := json.Marshal(newBinding)
	patchBytes, _ := jsonpatch.CreateMergePatch(oldData, newData)
	fmt.Println(string(patchBytes))
```

```json
{"status":{"initializers":null}}

```